### PR TITLE
Added a more descriptive error message for file operations in core_bind.cpp

### PR DESCRIPTION
**Pull request by Anutrix**
~Added a more descriptive error message for file operations in core_bind.cpp for android. Just a tip that it might because of lack of read-write permission.~

Added a more descriptive error message for file operations in `core_bind.cpp`. Just a tip that it might because of lack of read-write permission in any platform. Thx @Xrayez for https://github.com/godotengine/godot/pull/43076#issuecomment-718248873.

Addresses an issue(#42959) in #42719.

### CODE REVIEW

**Comment by Anutrix** [comment]
Build failed for macOS editor due to 
```Post job cleanup.
/usr/bin/tar --posix --use-compress-program zstd -T0 -cf cache.tzst -P -C /Users/runner/work/godot/godot --files-from manifest.txt
Warning: uploadChunk (start: 167772160, end: 201326591) failed: connect ETIMEDOUT 13.107.42.16:443
/Users/runner/work/_actions/actions/cache/v2/dist/save/index.js:3136
                        throw new Error(`Cache upload failed because file read failed with ${error.message}`);
                        ^

Error: Cache upload failed because file read failed with EBADF: bad file descriptor, read
    at ReadStream.<anonymous> (/Users/runner/work/_actions/actions/cache/v2/dist/save/index.js:3136:31)
    at ReadStream.emit (events.js:210:5)
    at internal/fs/streams.js:167:12
    at FSReqCallback.wrapper [as oncomplete] (fs.js:470:5)```

**Comment by Anutrix** [comment]
Can I get approval for this for hacktoberfest?

**Comment by bruvzg** [comment]
> How do I fix it?

It's CI cache issue, I have restarted the job.

File: core/bind/core_bind.cpp
Comment (dreamsComeTrue) [inline core/bind/core_bind.cpp:1255]: Small optimization hint: can't it be extracted as a `const String` at the top of the file to avoid so many String duplicates?

Comment (Anutrix) [reply core/bind/core_bind.cpp:1255]: Can be done but I wasn't sure if I should do it. A  dev reading it would need to scroll to beginning of the file to look for it. I'll make the changes if that's fine.

Comment (Anutrix) [reply core/bind/core_bind.cpp:1255]: We can call the constant string `android_hint` I guess.

**Comment by akien-mga** [comment]
I'm not fond of adding compiler defines to a core, cross-platform API specifically for Android. I don't think the change to the error message improves usability significantly enough for that (especially knowing that you need to go through `adb logcat` logs to even access such errors on Android).

If there's a need for better documentation on what kind of permissions are needed to use the File API, this should go in the File API's documentation, not in error messages.

**Comment by Xrayez** [comment]
> Just a tip that it might because of lack of read-write permission.

I'd merge the two messages into a general one, while still providing a hint as to what kind of a problem may occur, and which may as well apply to other platforms, something like this:

*"File must be opened before use, or lacking read-write permission."*

**Comment by Anutrix** [comment]
Changed message according Xrayez's comment above. Now it applies to all platforms and doesn't use compiler defines.

**Review by Calinou — APPROVED** [review]

**Comment by reduz** [comment]
This looks good to me too, but it needs a rebase. That said, I think @mhilbrunner is working on improving the error reporting, so it may be better to hold off until this is done?

**Comment by mhilbrunner** [comment]
I see no issue with this, if this is rebased, it can be merged (should however say `File must be opened before use, or is lacking read-write permission`, note the **is**).

If it isn't rebased in time I'll just integrate it into the error work I'm doing, noting it down :)

**Comment by Anutrix** [comment]
@mhilbrunner Sry about the (super) long delay.

Rebased and fixed the typo.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
